### DISCUSSION TITLE
Using OpenMDAO for gradient-based shape optimization

### DISCUSSION THREAD

**Discussion by KapilKhanal** [capytaine/capytaine#175] [Q&A]
Hi, I apologize for the poorly formed question but I am trying to understand which part of the source code of Capytaine I can get the implicit states of the state variable.  I am trying to investigate more about the suggestion I got about PDE solvers : "Most of the PDE solvers have one method for implicit states convergence and one method for calculating functionals based on the state variables". I assume GMRES solver is the one that is solving the linear systems taking in Matrices  `K` and the boundary condition of the problem. However, the implicit variables and the residual function are where I am confused.  What will be the residual function in Capytaine source code?

**Comment by mancellin** [comment]
I'm not sure to understand the terms of the question. Could you provide some context on this quote? Did the author give some concrete example of PDE solver?

**Reply by KapilKhanal** [reply]
@mancellin the suggestions are from the SO question I asked. The bigger context is that I am trying to implement Capytaine in  openMDAO components for my thesis. Individually I have a pretty good grasp on both Capytaine and openMDAO but integrating them is very hard to understand.  SO question:  https://stackoverflow.com/questions/72875606/wrapping-an-external-solver-as-an-component-in-openmdao-and-how-to-specify-pytho/72928563#72928563.  Thank you.

**Reply by KapilKhanal** [reply]
"implement" is not the right word..but trying to integrate them so that I can couple capytaine with [other solvers](https://github.com/openmdao/mphys) openMDAO provides

**Reply by mancellin** [reply]
If I understand correctly:
You have to solve an optimization problem and running Capytaine is part of your cost function. You would like to solve this optimization problem using gradient-descent, thus you need the adjoint (the gradient) of Capytaine's solver. OpenMDAO can compute this adjoint, assuming you rewrite the solver using its primitive building blocks (because OpenMDAO knows how to compute the adjoints of these building blocks and how to combine them using the chain rule).

In the SO question, @JustinSGray gives an example of a simple solver rewritten using two of OpenMDAO's building blocks.
Indeed Capytaine involves solving a linear system ("a method for implicit states convergence") and then some matrix-vector products to deduce the desired outputs ("a method for calculating functionals based on the state variables"). But it does not exactly fits in this nomenclature. In particular, another critical and costly step is the computation of these matrices (in the Fortran core) and computing the adjoint of this step might require some efforts.

**Reply by KapilKhanal** [reply]
Hi yes, that is the gist. If we could wrap capytaine in openMDAO, then it can be used with other solvers as well. For the Fortran core, is it possible to implement that in something like numpy ? or should we symbolically derive it and implement it on numpy? Also is there any method you've seen other researcher use capytaine for gradient based shape optimization?

**Reply by mancellin** [reply]
> For the Fortran core, is it possible to implement that in something like numpy ?

Yes, but expect a big loss in performance.

> or should we symbolically derive it and implement it on numpy?

In infinite depth, it might not be so difficult to implement an efficient analytical evaluation of the gradient based on the existing routines. But still a fair amount of work. Depends what kind of thesis you are working on :)

> Also is there any method you've seen other researcher use capytaine for gradient based shape optimization?

I believe @marcogambarini is aiming towards something like that, but his work is still early. Otherwise, no, I'm not aware of any other work of this kind.

**Comment by mancellin** [comment]
Capytaine might not fit exactly in this nomenclature.
FYI, here is a simplified diagram of the internals of Capytaine's solver, including intermediate internal variables.

```mermaid
flowchart TD;
    h[Water depth] --> gf(Assembling matrices);
    ω[Wave frequency] --> gf(Assembling matrices);
    m[Mesh] --> gf;
    gf -- K matrix --> ls(Linear solver);
    un[Normal velocity on hull] --> ls;
    gf -- S matrix --> mvp(Matrix vector product);
    ls -- sources distribution --> mvp;
    mvp -- potential distribution --> int("Integrate on mesh\n(basically a dot product)");
    m --> int;
    int --> f["Hydrodynamic forces\n(aka added mass and radiation damping)"]

    classDef input fill:#FFAAAA,color:#550000,stroke:#113939
    classDef step fill:#88BBBB,color:#003333,stroke:#226666
    classDef output fill:#FFE3AA,color:#553900,stroke:#AA8439
    class ω,m,un,h input
    class gf,ls,mvp,int step
    class f output
```

**Reply by KapilKhanal** [reply]
Awesome, thank you for sharing this flowchart. May I ask what DSL are you using to create this chart?

**Reply by mancellin** [reply]
It is https://mermaid-js.github.io

**Reply by JustinSGray** [reply]
Depending on what you're trying to do, you may or may not have to differentiate through the creation of the matrix. If the opt you are trying to achieve has fixed wave conditions, and the mesh "topology" is fixed (it normally is in this context) then you only have to account for the movement of the mesh points themselves. 

I don't know enough about this PDE solver to be sure, of course, but typically if the number and connectivity of the mesh nodes (i.e. its topology) is fixed then the structure of the matrix is fixed. Only the values that go into that matrix change. In that case the AD is normally much much easier, since you don't have to account for the matrix assembly process. You just take that as a given, and differentiate through the calculation of the matrix-values themselves. 

Im not saying its easy ... :) Just saying that its potentially not as hard as it sounds.

**Reply by mancellin** [reply]
Good idea, keeping the mesh fixed and changing only the velocity on the hull might be easier.

However, "mesh" here denotes not only the topology but also the mean position of the body. The "velocity" is only a small motion or deformation around this mean position.

To do shape optimisation, I guess you have in mind changing the mean position. But it could be a good start to consider a simpler problem such as optimizing the shape of the small deformations of a flexible body around a fixed mean position.

**Reply by KapilKhanal** [reply]
@mancellin Thank you for this very helpful discussion. I agree that I should first tackle the simpler problem. However, for my thesis, I might need the adjoint implemented but starting small is the way to go.  Also, could you explain what you meant by the fixed mean position? Position as in the position of the floating body or the position of the mesh nodes? similarly, I am a bit confused about the 'velocity on the hull 'as well.  Thank you!

**Reply by KapilKhanal** [reply]
@JustinSGray  Thank you very much for this helpful thread. I was originally planning to re-mesh every time the dimensions of my floating body change but since that will change the mesh nodes and connectivity of it, I will look into mesh morphing. @mancellin do you think the `body.keep_immersed_part()` is going to be an issue in ensuring the mesh nodes and connectivity same so that we get the same matrix structure?  @JustinSGray I am a little confused by what `account for mesh movements' means in the gradient calculations? Thank you both so much for engaging in this discussion. This helps me a lot!

**Reply by JustinSGray** [reply]
Mesh morphing isn't a super simple topic that I can easily summarize here. The best thing I can do practically is point you to the UofM MDO Lab to read some of [their publications on the subject](https://mdolab.engin.umich.edu/publications).  They also have some [open source python/Fortran](https://github.com/mdolab/pygeo) codes that do mesh morphing well (and with derivatives) (and OpenMDAO wrappers for them in [Mphys ](https://github.com/OpenMDAO/mphys). I really suggest you dig into the MDOLab work and Mphys if you're going down this route... 

but I will try to breifly explain the idea of mesh morphing. 

Broadly speaking for a PDE solver (CFD/FEA) you can think of the physical geometry you care about as a surface mesh. Then you have to create a volume mesh around that, which of course is itself dependent on the surface mesh. When you want to change the geometry, you are changing the surface mesh --- which then affects the volume mesh. How do you change the surface mesh? 

In reality, you have some set of physical parameters like wing-span, aspect-ratio, tower diameter. You then have some kind of function that maps those to a surface mesh. That function could be a cad program, or some spline-surface engine (like OpenVSP). 
When you set new parameter values, many people will think about generating a whole new (and usually different) surface mesh. Basically re-discritizing the surface geometry for every new parameter. However, for gradient based methods this approach is flawed. Re-discritizing introduces a non differentiable operation into the compute chain and you want to avoid that. 

So instead, we seek a method that can map geometric parameter changes to a constant surface mesh topology. In other words, we hold the number of mesh nodes and their connectivity fixed and only move the mesh points around in space. 

In some cases, all you have is the baseline surface mesh, in which case a free-form-deformation method can be used to create design variables that can pinch/pull the surface mesh to affect geometry changes. In this method, you get constant surface mesh topology for free. If you want to use CAD or something like OpenVSP, you have to set up a way to map their geometric changes to a constant mesh. Its not easy. Don't re-invent it. Look at what pyGeo does with OpenVSP. 

So assuming you've achieved a constant surface mesh topology, you now need to deal with the volume mesh. As the surface mesh changes, you have to keep the same volume mesh topology but morph the volume point locations to adjust to the new surface. Again, this is a complex topic. Lots of ways to do it. The one I recommend it the inverse distance weighting method that UofM implements in their tools. Start there. Don't try to write your own! 

So you now have a smooth, differentiable way to change your geometry: x -> surface_mesh(x) -> volume_mesh(surface_mesh(x)) -> pde_solver(volume_mesh) -> functionals(pde_states)

Now you have to differentiate that whole chain. The PDE solver is where the adjoint gets done, but you need derivatives of all those steps. Geometry, mesh morphing, solver, functionals. Then you chain them all together to do optimization.

**Reply by KapilKhanal** [reply]
@JustinSGray My understanding was that Pygeo and DVGeomtry are tailored for airfoils? Can it be used for any surfaces?

**Reply by JustinSGray** [reply]
can absolutely be used for any surfaces. The FFD's are totally general to any mesh. Nothing is really airfoil specific that I can think of anywhere in there. The CFD solver they use, ADflow... sure. But not the geometry stuff.

**Reply by KapilKhanal** [reply]
@JustinSGray This is assuring because I started learning about how MACH-Aero does it but then most of the geometric transformations ( like shear, taper, etc) methods are very airfoil specific, and also DVConstraints, generating FFD, etc were very 'Wing' specific. But, it seems I should have done proper diligence on what the pygeo 'engine' can do instead of the public-facing methods that are available. I will look into it and try to implement methods for the kinds of transformation I need. For my thesis, it would be more like geometry x: surface(x)-->BEMSolver(x)-->functionals(hydrodynamic coefficients). This is exciting! Thank you for this discussion!

**Reply by mancellin** [reply]
>  Also, could you explain what you meant by the fixed mean position? Position as in the position of the floating body or the position of the mesh nodes?

Position of the mesh nodes, and thus position of the floating body.

In frequency domain, the body motion can be seen as the oscillation of each node (in pratice each face center) around a mean position.
Linear potential flow theory is derived by assuming small body motion, so the oscillating motion is assumed to be along a straight line.
What I called "velocity on hull", that is also refered to as "dof" in the code, is the direction and amplitude of the oscillation on each face.

> do you think the body.keep_immersed_part() is going to be an issue in ensuring the mesh nodes and connectivity same so that we get the same matrix structure?

It could. It would be safer to design your parametrised geometry `x -> surface(x)` in such a way that all nodes are underwater.

**Reply by KapilKhanal** [reply]
@mancellin  wouldn't moving mesh nodes willy-nilly by an optimizer create an open boundary and mesh not watertight? Do you have example script that can create a 'bump' on the surface of a floating body? Any resources on using Capytaine for flexible dofs?  Thank you!

**Reply by mancellin** [reply]
> wouldn't moving mesh nodes willy-nilly by an optimizer create an open boundary and mesh not watertight?

Indeed, I would not recommend moving each node individually. I'm not sure about the risk of an open boundary, but you might just end up with a very "noisy" mesh that disturbs the solver. I have no experience with shape optimization, but I guess you should parametrize your shape with a few parameters (a simple example would be the principal axes of an ellipsoid) and optimize these parameters.

> Do you have example script that can create a 'bump' on the surface of a floating body?

```
import capytaine as cpt
body = cpt.Sphere(axial_symmetry=False)
body.mesh.triangulate_quadrangles()
vertices = body.mesh.vertices
vertices[4, :] += [-0.2, 0.0, 0.0]
body.mesh.vertices = vertices
body.show()
```
but I would strongly recommend using a dedicated meshing tool and trying to keep the mesh smooth.

> Any resources on using Capytaine for flexible dofs?

Unfortunately, this feature is under-documented. But look at the "Custom dof" example in the Cookbook in the documentation.
